codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -69,61 +70,183 @@
     private final Http2Settings initialSettings;
     private ChannelFutureListener closeListener;
     private BaseDecoder byteDecoder;
-    private long gracefulShutdownTimeoutMillis = DEFAULT_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS;
+    private long gracefulShutdownTimeoutMillis;
 
-    public Http2ConnectionHandler(boolean server) {
-        this(server, true);
+    /**
+     * Builder which builds {@link Http2ConnectionHandler} objects.
+     */
+    public static final class Builder extends BuilderBase<Http2ConnectionHandler, Builder> {
+        @Override
+        public Http2ConnectionHandler build0(Http2ConnectionDecoder decoder,
+                                             Http2ConnectionEncoder encoder) {
+            return new Http2ConnectionHandler(decoder, encoder, initialSettings());
+        }
     }
 
-    public Http2ConnectionHandler(boolean server, boolean validateHeaders) {
-        this(new DefaultHttp2Connection(server), validateHeaders);
-    }
+    /**
+     * Base class for a {@code builder} of any subclass of {@link Http2ConnectionHandler}.
+     * @param <T> The type of handler created by this builder.
+     * @param <B> The concrete type for this builder.
+     */
+    public abstract static class BuilderBase<T extends Http2ConnectionHandler, B extends BuilderBase<T, B>> {
+        private Http2Settings initialSettings = new Http2Settings();
+        private Http2FrameListener frameListener;
+        private Http2FrameLogger frameLogger;
+        private boolean validateHttp2Headers = true;
```

Comment:
Would we ever want to validate only one of http or http2 headers? If not, then maybe we should just control validation with a single flag (e.g. `validateHeaders`).